tests/channel-qvm-tests.lisp

```diff
@@ -0,0 +1,180 @@
+         (rule1 (qvm::make-noise-rule p1 kraus))
+         (rule2 (qvm::make-noise-rule p2 kraus))
+         (test-gate-1 (quil::build-gate "X" () 0))
+         (test-gate-2 (quil::build-gate "CNOT" () 0 1)))
+    (is (qvm::rule-matches-instr-p rule1 test-gate-1 posn))
+    (is (not (qvm::rule-matches-instr-p rule1 test-gate-2 posn)))
+    (is (qvm::rule-matches-instr-p rule2 test-gate-2 posn))
+    (is (not (qvm::rule-matches-instr-p rule2 test-gate-1 posn)))))
+
+
+(deftest test-match-rules ()
+  ;; Test that MATCH-RULES correctly returns the kraus operators for
+  ;; the matching rule of the instruction.
+  (let* ((posn :after)
+         (p1 (qvm::make-noise-pred (qvm::match-all-nq-gates 1) 12 posn))
+         (p2 (qvm::make-noise-pred (qvm::match-strict-qubits 1 2) 18 posn))
+         (kraus1 (qvm::generate-damping-kraus-map 5 1))
+         (kraus2 (qvm::generate-damping-kraus-map 6 2))
+         (rule1 (qvm::make-noise-rule p1 kraus1))
+         (rule2 (qvm::make-noise-rule p2 kraus2))
+         (rules (list rule1 rule2))
+         (test-gate-1 (quil::build-gate "X" () 0))
+         (test-gate-2 (quil::build-gate "CNOT" () 1 2))
+         (test-gate-3 (quil::build-gate "CNOT" () 3 4))
+         (rule-1-match (qvm::match-rules rules test-gate-1 posn)) 
+         (rule-2-match (qvm::match-rules rules test-gate-2 posn)))
+    ;; Check that the kraus operators returned (gate-x-match) are correct. 
+    (is (every #'cl-quil::matrix-equality (first (qvm::operation-elements rule-1-match)) kraus1))
+    (is (every #'cl-quil::matrix-equality (first (qvm::operation-elements rule-2-match)) kraus2))
+    (is (qvm::match-rules rules test-gate-2 posn))
+    (is (not (qvm::match-rules rules test-gate-3 posn)))
+    ))
+
+
+(deftest test-check-povm ()
+  ;; Test that check-povm rejects invalid povms, and accepts valid
+  ;; ones.
+  (let ((valid-povm (list .9d0 .1d0 .1d0 .9d0))
+        (invalid-povm1 (list .9d0 .1d0 .2d0 .9d0))
+        (invalid-povm2 (list .9 .1 .1 .2)))
+    (qvm::check-povm valid-povm)
+    (signals error (qvm::check-povm invalid-povm1))
+    (signals error (qvm::check-povm invalid-povm2))))
+
+
+(deftest test-check-kraus-ops ()
+  ;; Test that CHECK-KRAUS-OPS rejects invalid kraus ops and accepts
+  ;; valid ones.
+  (let ((valid-k1 (qvm::generate-damping-kraus-map 5 1))
+        (valid-k2 (qvm::generate-dephasing-kraus-map 5 2))
+        (invalid-k1 (qvm::generate-damping-kraus-map 2 5)) ; t1 < gate-time 
+        (invalid-k2 '(4 3 2 1))) ; not a magicl matrix
+    (qvm::check-kraus-ops valid-k1)
+    (qvm::check-kraus-ops valid-k2)
+    (signals error (qvm::check-kraus-ops invalid-k1))
+    (signals error (qvm::check-kraus-ops invalid-k2))))
+
+
+(deftest test-channel-qvm-noisy-readout ()
+  ;; Test that noisy readout works on the channel qvm. Applying a
+  ;; program 100 times and evaluate that the resulting excited state
+  ;; population is affected by the reaodut noise.
+  (let* ((nm (make-instance 'qvm::noise-model))
```

Comment:
The reason I added it in the approx-qvm tests was because somehow the program was being compiled before `run`. I think I solved the issue with the addition of some functions into `channel-qvm` from `noisy-qvm` to prevent programs from being compiled for channel/approx qvms and that seems to have worked. I think I'll add the guard in channel qvm as a safeguard as well.